File: explore/vehicle_positions.py
#!/usr/bin/env -S uv run --script
# /// script
# requires-python = ">=3.9"
# dependencies = []
# ///
"""Show live ferry positions from the GTFS-rt vehicle-positions feed.

    uv run vehicle_positions.py                  # all vehicles
    uv run vehicle_positions.py --route 19417    # one route only

Useful later for a "ferry is underway" indicator on the TRMNL display.
"""

import argparse
import time

from swiftly import AGENCY_KEY, get_json, pick


def vehicles(feed):
    """Yield dicts describing each vehicle in the feed."""
    for entity in feed.get("entity", []):
        v = pick(entity, "vehicle")
        if not v:
            continue
        trip = pick(v, "trip") or {}
        pos = pick(v, "position") or {}
        inner = pick(v, "vehicle") or {}
        ts = pick(v, "timestamp")
        yield {
            "vehicle_id": str(pick(inner, "id") or pick(entity, "id") or ""),
            "label": pick(inner, "label") or "",
            "route_id": str(pick(trip, "routeId", "route_id") or ""),
            "lat": pick(pos, "latitude", "lat"),
            "lon": pick(pos, "longitude", "lon"),
            "speed": pick(pos, "speed"),
            "status": pick(v, "currentStatus", "current_status") or "",
            "stop_id": str(pick(v, "stopId", "stop_id") or ""),
            "timestamp": int(ts) if ts else None,
        }


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--route", help="filter to this route id")
    args = parser.parse_args()

    feed = get_json(f"/real-time/{AGENCY_KEY}/gtfs-rt-vehicle-positions", format="json")
    now = int(time.time())

    rows = [v for v in vehicles(feed) if not args.route or v["route_id"] == args.route]
    if not rows:
        print("No vehicles match the filters.")
        return

    print(f"{'vehicle':<10} {'route':<8} {'position':<24} {'speed':<8} {'status':<14} age")
    for v in sorted(rows, key=lambda x: x["vehicle_id"]):
        loc = f"{v['lat']}, {v['lon']}" if v["lat"] is not None else "(no fix)"
        speed = f"{v['speed']:.1f}" if isinstance(v["speed"], (int, float)) else "-"
        age = f"{(now - v['timestamp'])}s" if v["timestamp"] else "-"
        status = v["status"]
        if v["stop_id"]:
            status = f"{status}@{v['stop_id']}"
        print(f"{v['vehicle_id']:<10} {v['route_id']:<8} {loc:<24} {speed:<8} {status:<14} {age}")


if __name__ == "__main__":
    main()

File: scripts/vehicle-positions.mjs
#!/usr/bin/env node
// Show live ferry positions from the GTFS-rt vehicle-positions feed.
//
//   node scripts/vehicle-positions.mjs                  # all vehicles
//   node scripts/vehicle-positions.mjs --route 19417    # one route only
//
// Useful later for a "ferry is underway" indicator on the TRMNL display.

import { parseArgs } from 'node:util';
import { AGENCY_KEY, getJson, pick } from './swiftly.mjs';

const { values } = parseArgs({ options: { route: { type: 'string' } } });

// Yield objects describing each vehicle in the feed.
function* vehicles(feed) {
  for (const entity of feed.entity ?? []) {
    const v = pick(entity, 'vehicle');
    if (!v) continue;
    const trip = pick(v, 'trip') ?? {};
    const pos = pick(v, 'position') ?? {};
    const inner = pick(v, 'vehicle') ?? {};
    const ts = pick(v, 'timestamp');
    yield {
      vehicleId: String(pick(inner, 'id') ?? pick(entity, 'id') ?? ''),
      label: pick(inner, 'label') ?? '',
      routeId: String(pick(trip, 'routeId', 'route_id') ?? ''),
      lat: pick(pos, 'latitude', 'lat'),
      lon: pick(pos, 'longitude', 'lon'),
      speed: pick(pos, 'speed'),
      status: pick(v, 'currentStatus', 'current_status') ?? '',
      stopId: String(pick(v, 'stopId', 'stop_id') ?? ''),
      timestamp: ts ? Number(ts) : null,
    };
  }
}

const feed = await getJson(`/real-time/${AGENCY_KEY}/gtfs-rt-vehicle-positions`, {
  format: 'json',
});
const now = Math.floor(Date.now() / 1000);

const rows = [...vehicles(feed)].filter((v) => !values.route || v.routeId === values.route);
if (rows.length === 0) {
  console.log('No vehicles match the filters.');
  process.exit(0);
}

console.log(
  `${'vehicle'.padEnd(10)} ${'route'.padEnd(8)} ${'position'.padEnd(24)} ${'speed'.padEnd(8)} ${'status'.padEnd(14)} age`,
);
for (const v of rows.sort((a, b) => a.vehicleId.localeCompare(b.vehicleId))) {
  const loc = v.lat != null ? `${v.lat}, ${v.lon}` : '(no fix)';
  const speed = typeof v.speed === 'number' ? v.speed.toFixed(1) : '-';
  const age = v.timestamp ? `${now - v.timestamp}s` : '-';
  const status = v.stopId ? `${v.status}@${v.stopId}` : v.status;
  console.log(
    `${v.vehicleId.padEnd(10)} ${v.routeId.padEnd(8)} ${loc.padEnd(24)} ${speed.padEnd(8)} ${status.padEnd(14)} ${age}`,
  );
}
